### PR TITLE
Add basic input configuration

### PR DESCRIPTION
**Pull request by Leon-Plickat**
Since our original idea of doing this via a Wayland protocol will probably not happen in the near future and since that one hack to configure libinput isn't very pretty I decided to experiment with simple input configuration in river itself. No idea if this will lead anywhere...

Keyboard configuration is deliberately left out for now, this work is focusing on libinput.

#### TODO

* [x] Implement `wlr.InputDevice.getLibinputDeviceHandle()`
* [x] Implement ~~all the common~~ *a lot of the* libinput configuration things
* [x] Add a command to add one configuration to one input device, matching by identifier
* [X] Apply to new devices
* [x] Apply to existing devices

### CODE REVIEW

File: doc/riverctl.1.scd
Comment (ifreund) [inline doc/riverctl.1.scd:276]:
```suggestion
A list of all device properties that can be configured may be found below. However note
```

File: doc/riverctl.1.scd
Comment (ifreund) [inline doc/riverctl.1.scd:279]: What does `disabled-on-external-mouse` mean? Should we perhaps point users to libinput docs on these options?

File: river/InputManager.zig
Comment (ifreund) [inline river/InputManager.zig:55]: https://github.com/ziglang/zig/pull/8698
```suggestion
                mem.sliceTo(device.name, 0),
```

File: river/InputManager.zig
Comment (ifreund) [inline river/InputManager.zig:50]:
```suggestion
        const identifier = try std.fmt.allocPrint(
```

File: river/InputManager.zig
Comment (ifreund) [inline river/InputManager.zig:45]: should be a `///` doc comment

File: river/command/input.zig
Comment (ifreund) [inline river/command/input.zig:51]: Instead of making `input_config` optional and checking if it is null here, it would be more idiomatic to use the else clause of the while loop to handle the case in which it does not exist:
```
const input_config = while (...) {
} else blk: {
     // create input config
     break :blk input_config;
};
```
with this approach the `.?` on `input_config` throughout the rest of this function can be avoided.

File: river/command/input.zig
Comment (ifreund) [inline river/command/input.zig:99]:
```suggestion
    // Update matching existing input devices
```

File: river/InputConfig.zig
Comment (ifreund) [inline river/InputConfig.zig:43]: If we used `@"disabled-on-external-mouse"` instead the parse function could be replaced by `std.meta.stringToEnum()`, which would be cleaner IMO. The same goes for all the other enums in this file.

File: river/InputConfig.zig
Comment (ifreund) [inline river/InputConfig.zig:66]: Not a hard rule since I don't have any way to enforce it, but the style I've been moving towards for if statements is no braces if they fit on one line but always braces if they don't:
```zig
if (foo) bar(); // good

if (foo) {
    bar(); // good
}

if (foo)
    bar(); // I don't like this
```

**Review by ifreund — COMMENTED** [review]
Here's some initial code review, I haven't done any actual testing yet.

File: doc/riverctl.1.scd
Comment (novakne) [reply doc/riverctl.1.scd:279]: it means 'disable the device while an external mouse is plugged in', not sure if there is a man page instead of https://wayland.freedesktop.org/libinput/doc/latest/configuration.html

**Comment by Leon-Plickat** [comment]
@ifreund  Your first round of reviews managed to knock off about 200 lines!

File: doc/riverctl.1.scd
Comment (ifreund) [inline doc/riverctl.1.scd:274]: How does one find out the device names? I'm assuming there's some command to list them but the average user probably isn't aware of it.

File: river/command/input.zig
Comment (ifreund) [inline river/command/input.zig:118]: We could avoid this check (and the existence of this wrapper function) by removing the `unset` field from the enums and making the fields of `InputConfig` `?EnumFoo` instead, defaulting to `null`.

File: doc/riverctl.1.scd
Comment (Leon-Plickat) [reply doc/riverctl.1.scd:274]: There isn't a good way currently. The best way is either to read the logs for "new input device" messages, or to hop into sway and use `swaymsg -t get_inputs`.

Comment (ifreund) [reply doc/riverctl.1.scd:274]: looks like libinput ships a command line tool that can provide this information. `libinput list-devices` prints off a list for me, although the names aren't in the format we expect.

Comment (Leon-Plickat) [reply doc/riverctl.1.scd:274]: I decided to use the same scheme to build the identifiers as Sway does (`<vendor id>:<product id>:<name>`) for the sake of consistency.

The identifier sadly can't be just the advertised device name, as that would lead to even more duplicate identifiers.

`libinput list-devices` does not show the vendor or device id unfortunately.

Comment (novakne) [reply doc/riverctl.1.scd:274]: Would it be hard to implement a command like `riverctl devices-info` that return the devices identifier, maybe using json? Could be a better user friendly way than looking at the log

File: river/command/input.zig
Comment (ifreund) [inline river/command/input.zig:143]: We need an `else` branch here to error on an unknown subcommand.

File: doc/riverctl.1.scd
Comment (ifreund) [inline doc/riverctl.1.scd:334]: can we put this first in the input section to make it a bit more discoverable?

**Review by ifreund — COMMENTED** [review]
I hit a panic while testing:
```
thread 3392 panic: invalid enum value
???:?:?: 0x294216 in ??? (???)
/home/ifreund/projects/river/river/command/input.zig:40:36: 0x21b291 in command.input.listInputs (river)
        try input_list.appendSlice(@tagName(node.data.device.type));
                                   ^
/home/ifreund/projects/river/river/command.zig:128:16: 0x2538f1 in command.run (river)
    try impl_fn(allocator, seat, args, out);
               ^
/home/ifreund/projects/river/river/Control.zig:107:24: 0x23e55f in Control.handleRequest (river)
            command.run(util.gpa, seat, args.items, &out) catch |err| {
                       ^
/home/ifreund/projects/river/zig-cache/zig-wayland/common.zig:101:76: 0x2546e9 in .wayland.common.Dispatcher(.wayland.river_control_unstable_v1_server.ControlV1,*Control).dispatcher (river)
                    @ptrCast(fn (*Obj, Payload, Data) void, implementation)(
                                                                           ^
```

Also, we should probably add completions for these new riverctl subcommands...

**Comment by Leon-Plickat** [comment]
For that panic to happen, I assume `wlr.InputDevice.type` was set to a value not in `wlr.InputDevice.Type`. Since that struct is `extern` the problem may not be on our end.

**Comment by Leon-Plickat** [comment]
Did you unplug the mouse and plugged it back in again? I think I forgot to handle input device removals.

**Comment by Leon-Plickat** [comment]
Removed devices are now handled. That panic should no longer happen.

File: river/InputManager.zig
Comment (ifreund) [inline river/InputManager.zig:98]: We don't seem to store pointers to `InputConfig`s anywhere which would require pointer stability. Therefore, I think we could just use an `ArrayList` (or `ArrayListUnmanaged`) here which should simplify things a bit.

**Review by ifreund — COMMENTED** [review]
Nice, that seems to have fixed things as I can no longer reproduce the panic. Just one more simplification I noticed then I think this is good to merge!

**Comment by Leon-Plickat** [comment]
I wonder if we can use an ArrayList for `InputManager.input_devices` as well.

**Comment by Leon-Plickat** [comment]
I also decided to add a `list-input-configs` command, which I found useful for testing.

File: river/command/input.zig
Comment (ifreund) [inline river/command/input.zig]: I'd do the following as the control flow is a bit simpler.
```zig
try server.input_manager.input_configs.ensureUnusedCapacity(1);
server.input_manager.input_configs.appendAssumeCapacity(.{ .identifier = try util.gpa.dupe(u8, args([1]) });
```

File: river/InputManager.zig
Comment (ifreund) [inline river/InputManager.zig:98]: We should not use a default value here, as omitting `input_configs` when initializing `InputManager` would be a bug.

File: river/command/input.zig
Comment (ifreund) [inline river/command/input.zig:72]: We can use the writer interface and std.fmt to make this code less verbose. 
```suggestion
            try writer.print("\tevents: {}\n", .{event_state});
```

File: river/command/input.zig
Comment (ifreund) [inline river/command/input.zig:63]:
```suggestion
    var input_list = std.ArrayList(u8).init(allocator);
    const writer = input_list.writer();
```

**Review by ifreund — COMMENTED** [review]
> I wonder if we can use an ArrayList for InputManager.input_devices as well.

We can't as we need the input devices to be pinned in memory due to the destroy listener.

> I also decided to add a list-input-configs command, which I found useful for testing.

Nice, this is certianly useful for testing and likely for users as well :)

**Comment by ifreund** [comment]
Oh, and the new commands are missing from the completion scripts currently. It would be nice to add them but I won't block merging this on that. (It would block 0.1.0 though)

File: river/command/input.zig
Comment (ifreund) [inline river/command/input.zig:73]: Ok, I promise this is my last bikeshed before merging this :D
```suggestion

    for (server.input_manager.input_configs.items) |*input_config, i| {
        if (i > 0) try input_list.appendSlice("\n");
```

**Review by ifreund — APPROVED** [review]
Great work on this, LGTM!

**Pull request merged**
